### PR TITLE
[new release] tyxml-lwd, nottui, nottui-pretty, nottui-lwt, lwd and brr-lwd (0.2)

### PR DESCRIPTION
**Pull request by let-def**
Make reactive webpages in Js_of_ocaml using Tyxml and Lwd

- Project page: <a href="https://github.com/let-def/lwd">https://github.com/let-def/lwd</a>
- Documentation: <a href="https://let-def.github.io/lwd/doc">https://let-def.github.io/lwd/doc</a>

##### CHANGES:

Sun Feb 20 20:49:47 JST 2022

- Lwd.fix operator helps working with graphs that cannot be evaluated in a
  single pass
- brr-lwd library integrates Lwd with Brr library, for writing javascript
  applications

Bug fixes:
- fixed invalidation in Lwd
- restored some internal invariants in Lwd_seq
- fixed behavior of Notty sensors

### CODE REVIEW

File: packages/nottui/nottui.0.2/opam [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+synopsis: "UI toolkit for the terminal built on top of Notty and Lwd"
+maintainer: ["fred@tarides.com"]
+authors: ["Frédéric Bour"]
+license: "MIT"
+homepage: "https://github.com/let-def/lwd"
+doc: "https://let-def.github.io/lwd/doc"
+bug-reports: "https://github.com/let-def/lwd/issues"
+depends: [
+  "dune" {>= "2.7" & >= "2.7"}
+  "lwd"
```
Comment (mseri) [inline packages/nottui/nottui.0.2/opam:11]: To fix the lower bound issues
```
# 21 |   let fmap f x = Lwd.map ~f x
#                                ^
# Error: The function applied to this argument has type 'a Lwd.t -> 'b Lwd.t
# This argument cannot be applied with label ~f
```
I think you need
```suggestion
  "lwd" {=version}
```
or, at least, `{>= "0.2"}`

File: packages/tyxml-lwd/tyxml-lwd.0.2/opam [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+synopsis: "Make reactive webpages in Js_of_ocaml using Tyxml and Lwd"
+maintainer: ["fred@tarides.com"]
+authors: ["Frédéric Bour"]
+license: "MIT"
+homepage: "https://github.com/let-def/lwd"
+doc: "https://let-def.github.io/lwd/doc"
+bug-reports: "https://github.com/let-def/lwd/issues"
+depends: [
+  "dune" {>= "2.7" & >= "2.7"}
+  "lwd"
```
Comment (mseri) [inline packages/tyxml-lwd/tyxml-lwd.0.2/opam:11]: To fix the lower bound issues
```
# 21 |   let fmap f x = Lwd.map ~f x
#                                ^
# Error: The function applied to this argument has type 'a Lwd.t -> 'b Lwd.t
# This argument cannot be applied with label ~f
```
I think you need
```suggestion
  "lwd" {=version}
```
or, at least, `{>= "0.2"}`

File: packages/tyxml-lwd/tyxml-lwd.0.2/opam [hunk 3]
```diff
@@ -0,0 +1,40 @@
+opam-version: "2.0"
+synopsis: "Make reactive webpages in Js_of_ocaml using Tyxml and Lwd"
+maintainer: ["fred@tarides.com"]
+authors: ["Frédéric Bour"]
+license: "MIT"
+homepage: "https://github.com/let-def/lwd"
+doc: "https://let-def.github.io/lwd/doc"
+bug-reports: "https://github.com/let-def/lwd/issues"
+depends: [
+  "dune" {>= "2.7" & >= "2.7"}
```
Comment (mseri) [inline packages/tyxml-lwd/tyxml-lwd.0.2/opam:10]:
```suggestion
  "dune" {>= "2.7"}
```
To fix this, you should remove `dune` from the dependencies in the dune-project file. It is added automatically

File: packages/nottui-pretty/nottui-pretty.0.2/opam [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+maintainer: ["fred@tarides.com"]
+authors: ["Frédéric Bour"]
+license: "MIT"
+homepage: "https://github.com/let-def/lwd"
+doc: "https://let-def.github.io/lwd/doc"
+bug-reports: "https://github.com/let-def/lwd/issues"
+depends: [
+  "dune" {>= "2.7" & >= "2.7"}
+  "notty"
+  "nottui"
```
Comment (mseri) [inline packages/nottui-pretty/nottui-pretty.0.2/opam:12]:
```suggestion
  "nottui" {= version}
```
see comment below

File: packages/nottui-pretty/nottui-pretty.0.2/opam [hunk 5]
```diff
@@ -0,0 +1,38 @@
+opam-version: "2.0"
+synopsis: "A pretty-printer based on PPrint rendering UIs"
+maintainer: ["fred@tarides.com"]
+authors: ["Frédéric Bour"]
+license: "MIT"
+homepage: "https://github.com/let-def/lwd"
+doc: "https://let-def.github.io/lwd/doc"
+bug-reports: "https://github.com/let-def/lwd/issues"
+depends: [
+  "dune" {>= "2.7" & >= "2.7"}
```
Comment (mseri) [inline packages/nottui-pretty/nottui-pretty.0.2/opam:10]:
```suggestion
  "dune" {>= "2.7"}
```
see the other comment

File: packages/nottui-lwt/nottui-lwt.0.2/opam [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+authors: ["Frédéric Bour"]
+license: "MIT"
+homepage: "https://github.com/let-def/lwd"
+doc: "https://let-def.github.io/lwd/doc"
+bug-reports: "https://github.com/let-def/lwd/issues"
+depends: [
+  "dune" {>= "2.7" & >= "2.7"}
+  "notty"
+  "lwt"
+  "nottui"
```
Comment (mseri) [inline packages/nottui-lwt/nottui-lwt.0.2/opam:13]:
```suggestion
  "nottui" {= version}
```
see other comment

File: packages/nottui-lwt/nottui-lwt.0.2/opam [hunk 7]
```diff
@@ -0,0 +1,39 @@
+opam-version: "2.0"
+synopsis: "Run Nottui UIs in Lwt"
+maintainer: ["fred@tarides.com"]
+authors: ["Frédéric Bour"]
+license: "MIT"
+homepage: "https://github.com/let-def/lwd"
+doc: "https://let-def.github.io/lwd/doc"
+bug-reports: "https://github.com/let-def/lwd/issues"
+depends: [
+  "dune" {>= "2.7" & >= "2.7"}
```
Comment (mseri) [inline packages/nottui-lwt/nottui-lwt.0.2/opam:10]:
```suggestion
  "dune" {>= "2.7"}
```

File: packages/lwd/lwd.0.2/opam [hunk 8]
```diff
@@ -0,0 +1,40 @@
+opam-version: "2.0"
+synopsis: "Lightweight reactive documents"
+maintainer: ["fred@tarides.com"]
+authors: ["Frédéric Bour"]
+license: "MIT"
+homepage: "https://github.com/let-def/lwd"
+doc: "https://let-def.github.io/lwd/doc"
+bug-reports: "https://github.com/let-def/lwd/issues"
+depends: [
+  "dune" {>= "2.7" & >= "2.7"}
```
Comment (mseri) [inline packages/lwd/lwd.0.2/opam:10]:
```suggestion
  "dune" {>= "2.7"}
```

File: packages/brr-lwd/brr-lwd.0.2/opam [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+synopsis: "Make reactive webpages in Js_of_ocaml using Brr and Lwd"
+maintainer: ["fred@tarides.com"]
+authors: ["Frédéric Bour"]
+license: "MIT"
+homepage: "https://github.com/let-def/lwd"
+doc: "https://let-def.github.io/lwd/doc"
+bug-reports: "https://github.com/let-def/lwd/issues"
+depends: [
+  "dune" {>= "2.7" & >= "2.7"}
+  "lwd"
```
Comment (mseri) [inline packages/brr-lwd/brr-lwd.0.2/opam:11]:
```suggestion
  "lwd" {= version}
```
see other comment

**Comment by mseri** [comment]
Tests are failing with:
```
#=== ERROR while compiling lwd.0.2 ============================================#
# context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.03.0 | pinned(https://github.com/let-def/lwd/releases/download/v0.2/nottui-lwt-0.2.tbz)
# path                 ~/.opam/4.03/.opam-switch/build/lwd.0.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p lwd -j 31 @install @runtest
# exit-code            1
# env-file             ~/.opam/log/lwd-1018-5f1023.env
# output-file          ~/.opam/log/lwd-1018-5f1023.out
### output ###
# File "lib/lwd/dune", line 1, characters 0-172:
# 1 | (library 
# 2 |   (name lwd) 
# 3 |   (public_name lwd)
# 4 |   (modules lwd lwd_seq lwd_table lwd_infix lwd_utils) 
# 5 |   (libraries seq)
# 6 |   (inline_tests (backend qtest.lib))
# 7 |   (wrapped false))
# (cd _build/default/lib/lwd && /home/opam/.opam/4.03/bin/qtest extract --quiet lwd_utils.ml lwd_table.ml lwd_seq.ml lwd_infix.ml lwd.ml lwd_utils.mli lwd_table.mli lwd_seq.mli lwd_infix.mli lwd.mli) > _build/default/lib/lwd/.lwd.inline-tests/inline_test_runner_lwd.ml-gen
# Fatal error: exception Sys_error("lwd_infix.ml: No such file or directory")
```

`nottui.0.1` is not compatible, needs to get an upper bound:
```
#=== ERROR while compiling nottui.0.1 =========================================#
# context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.13.1 | file:///home/opam/opam-repository
# path                 ~/.opam/4.13/.opam-switch/build/nottui.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p nottui -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/nottui-3053-e7966f.env
# output-file          ~/.opam/log/nottui-3053-e7966f.out
### output ###
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/nottui/.nottui.objs/byte -I /home/opam/.opam/4.13/lib/lwd -I /home/opam/.opam/4.13/lib/notty -I /home/opam/.opam/4.13/lib/seq -I /home/opam/.opam/4.13/lib/uchar -I /home/opam/.opam/4.13/lib/uucp -I /home/opam/.opam/4.13/lib/uuseg -I /home/opam/.opam/4.13/lib/uutf -intf-suffix .ml -no-alias-deps -o lib/nottui/.nottui.objs/byte/nottui.cmo -c -impl lib/nottui/nottui.ml)
# File "lib/nottui/nottui.ml", line 38, characters 8-15:
# 38 |     (v, Lwd.map (fun i -> Handle (i, v)) (Lwd.get v))
#              ^^^^^^^
# Warning 6 [labels-omitted]: label f was omitted in the application of this function.
# File "lib/nottui/nottui.ml", line 38, characters 16-40:
# 38 |     (v, Lwd.map (fun i -> Handle (i, v)) (Lwd.get v))
#                      ^^^^^^^^^^^^^^^^^^^^^^^^
# Error: This expression should not be a function, the expected type is
#        'a Lwd.t
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlopt.opt -w -40 -g -I lib/nottui/.nottui.objs/byte -I lib/nottui/.nottui.objs/native -I /home/opam/.opam/4.13/lib/lwd -I /home/opam/.opam/4.13/lib/notty -I /home/opam/.opam/4.13/lib/seq -I /home/opam/.opam/4.13/lib/uchar -I /home/opam/.opam/4.13/lib/uucp -I /home/opam/.opam/4.13/lib/uuseg -I /home/opam/.opam/4.13/lib/uutf -intf-suffix .ml -no-alias-deps -o lib/nottui/.nottui.objs/native/nottui.cmx -c -impl lib/nottui/nottui.ml)
# File "lib/nottui/nottui.ml", line 38, characters 8-15:
# 38 |     (v, Lwd.map (fun i -> Handle (i, v)) (Lwd.get v))
#              ^^^^^^^
# Warning 6 [labels-omitted]: label f was omitted in the application of this function.
# File "lib/nottui/nottui.ml", line 38, characters 16-40:
# 38 |     (v, Lwd.map (fun i -> Handle (i, v)) (Lwd.get v))
#                      ^^^^^^^^^^^^^^^^^^^^^^^^
# Error: This expression should not be a function, the expected type is
#        'a Lwd.t
# (cd _build/default && /home/opam/.opam/4.13/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/nottui/.nottui.objs/byte -I /home/opam/.opam/4.13/lib/lwd -I /home/opam/.opam/4.13/lib/notty -I /home/opam/.opam/4.13/lib/seq -I /home/opam/.opam/4.13/lib/uchar -I /home/opam/.opam/4.13/lib/uucp -I /home/opam/.opam/4.13/lib/uuseg -I /home/opam/.opam/4.13/lib/uutf -intf-suffix .ml -no-alias-deps -o lib/nottui/.nottui.objs/byte/nottui_widgets.cmo -c -impl lib/nottui/nottui_widgets.ml)
# File "lib/nottui/nottui_widgets.ml", line 97, characters 2-11:
# 97 |   Lwd.map2' t state @@ fun t state ->
#        ^^^^^^^^^
# Error: Unbound value Lwd.map2'
# Hint: Did you mean map2?
```

**Comment by let-def** [comment]
Many thanks for the review, I tried to update the description files. Let's see what we get this time :)

**Comment by kit-ty-kate** [comment]
@let-def does the current PR look ok to you? I’ve added `{= version}` because it was broken for the `0.1` packages and you seem to release all at once everytime so this makes it much easier for everyone to pack everything together.

If you prefer to be more precise you can switch to `{>= "0.2" & < "0.3"}` to allow patch versions on one package

**Comment by let-def** [comment]
Yes this looks good. Thanks for updating the other bounds!

**Comment by kit-ty-kate** [comment]
Thanks! Could you return the fixes in the opam files upstream?

**Comment by let-def** [comment]
The opam files generated by dune contain `"lwd" {= "version"}`.
Are the double-quotes changing the meaning?

**Comment by mseri** [comment]
Yes. You need to use `:version` in the dune project file to generate the correct value

**Comment by let-def** [comment]
Good to know, thanks.

**Pull request merged**